src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -0,0 +1,246 @@
+ *
+ * @template T of object
+ *
+ * @experimental
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class ObjectMapper implements ObjectMapperInterface
+{
+    public function __construct(
+        private readonly MapperMetadataFactoryInterface $metadataFactory = new ReflectionMapperMetadataFactory(),
+        private readonly ?PropertyAccessorInterface $propertyAccessor = null,
+    ) {
+    }
+
+    public function map(object $object, object|string|null $to = null): object
+    {
+        static $objectMap = null;
+        $objectMapInitialized = false;
+
+        if (null === $objectMap) {
+            $objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        try {
+            $refl = new \ReflectionClass($object);
+        } catch (\ReflectionException $e) {
+            throw new ReflectionException($e->getMessage(), $e->getCode(), $e);
+        }
+
+        $metadata = $this->metadataFactory->create($object);
+        $map = $this->getMapTo($metadata, null, $object);
+        $to ??= $map?->to;
+
+        if (!(\is_string($to) && class_exists($to)) && !\is_object($to)) {
+            throw new MappingException(sprintf('Attribute of type "%s" expected on "%s.', Map::class, $refl->getName()));
+        }
+
+        $mappingToObject = \is_object($to);
+        try {
+            $toRefl = new \ReflectionClass($to);
+        } catch (\ReflectionException $e) {
+            throw new ReflectionException($e->getMessage(), $e->getCode(), $e);
+        }
+
+        $mapped = $mappingToObject ? $to : $toRefl->newInstanceWithoutConstructor();
+        if ($map && $map->transform) {
+            $mapped = $this->applyTransforms($map, $mapped, $mapped);
+
+            if (!\is_object($mapped)) {
+                throw new MappingTransformException('Can not map to a non-object.');
+            }
+        }
+
+        if (!is_a($mapped, $toRefl->getName(), false)) {
+            throw new MappingException(sprintf('Expected the mapped object to be an instance of "%s".', $mappingToObject ? $to::class : $to));
+        }
+
+        $objectMap[$object] = $mapped;
+
+        $arguments = [];
+        $constructor = $toRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $parameterName = $parameter->getName();
+            if (!$toRefl->hasProperty($parameterName)) {
+                continue;
+            }
+
+            $property = $toRefl->getProperty($parameterName);
+
+            // The mapped class was probably instantiated in a transform we can't write a readonly property
+            if ($property->isReadOnly() && ($property->isInitialized($mapped) && $property->getValue($mapped))) {
+                continue;
+            }
+
+            $arguments[$parameterName] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
+        }
+
+        foreach ($refl->getProperties() as $property) {
```

Comment:
After some tests I think that this is not good enough as the metadata should provide a whole `source` => `target` tuple not only the target.